### PR TITLE
Adding ModelPart

### PR DESCRIPTION
**Pull request by philbucher**
This adds the `ModelPart` for the CoSimIO, which will serve as data-container esp for exchanging Mesh and (Geometry in the future)
I decided to follow the interface we have in Kratos to not overcomplicate / reinvent the wheel

To be discussed:
- [x] container-types: `private` / `public` ? => public
- [x] allowing non-const access? => probably not for now but maybe in future => will be necessary in future
- [x]  is using `std::size_t` ok? => thinking of MPI => typedef to `std::ptrdiff_t` (for reference: [long index discussion in Kratos](https://github.com/KratosMultiphysics/Kratos/pull/5361))
- [x] implement and add tests after discussing
- [x] `CoSimIO` namespace is missing :)
- [x] add `IndexType` and `CoordinatesType` to [define.h](https://github.com/KratosMultiphysics/CoSimIO/blob/master/co_sim_io/impl/define.hpp)

### CODE REVIEW

**Comment by philbucher** [comment]
more discussion points:
- [ ] `Element` stored `Node`s as `vector<Node*>` vs `vector<reference_wrapper<Node>>`. Using them is kinda nasty both times:
~~~cpp
// reference_wrapper<Node>
for (auto node_it=element.NodesBegin(); node_it!=element.NodesEnd(); ++node_it) {
    node_it->get().Id();
}

// Node*
for (auto node_it=element.NodesBegin(); node_it!=element.NodesEnd(); ++node_it) {
    (*node_it)->Id();
}
~~~

- [ ] Do we need the `Connectivities` / `NodeIds` function for `Element`? => might only become clearer when using it
- [ ] Having a `Nodes` function like `ModelPart` has it will require to expose the container right? (would still be covenient though :))

**Comment by philbucher** [comment]
@pooyan-dadvand I think this is ready. I would merge as is such that I can make the C and the Python interface. Once those are done then we are sure the basic capabilities work across languages and we can refine it  / start using it

for future PRs:
- [ ] switch from `shared_ptr` to `intrusive_ptr` (as discussed)
- [ ] use an `unordered_map` as additional data-structure to speed up the Id-based search

**Comment by pooyan-dadvand** [comment]
@philbucher I think you are missing the element type in the element and also in the CreateNewElement interface

File: co_sim_io/impl/model_part.hpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+class Element
+{
+public:
+    using ElementType = std::size_t;
+    using NodesContainerType = std::vector<Node*>;
+    using ConnectivitiesType = std::vector<IdType>;
+
+    Element(
+        const IdType I_Id,
+        const ElementType I_Type,
```
Comment (philbucher) [inline co_sim_io/impl/model_part.hpp:83]: it is here

File: co_sim_io/impl/model_part.hpp [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    {
+        CO_SIM_IO_ERROR_IF(HasNode(I_Id)) << "The Node with Id " << I_Id << " exists already!" << std::endl;
+
+        mNodes.push_back(std::make_shared<Node>(I_Id, I_X, I_Y, I_Z));
+        return *(mNodes.back());
+    }
+
+    Element& CreateNewElement(
+        const IdType I_Id,
+        const Element::ElementType I_Type,
```
Comment (philbucher) [inline co_sim_io/impl/model_part.hpp:146]: and here

**Review by philbucher — COMMENTED** [review]
> @philbucher I think you are missing the element type in the element and also in the CreateNewElement interface

no it is there

**Comment by pooyan-dadvand** [comment]
Let me go and prepare myself another coffee! Seems that I need it!

**Comment by pooyan-dadvand** [comment]
I just got confused between `IdType` and `I_Type`

**Review by pooyan-dadvand — APPROVED** [review]
Approving after a coffee! :+1:

**Pull request merged**